samples/net/http_server/src/main.c

```diff
@@ -0,0 +1,673 @@
+
+	if (IS_ENABLED(CONFIG_NET_SOCKETS_SOCKOPT_TLS)) {
+		*sock = socket(bind_addr->sa_family, SOCK_STREAM |
+						     SOCK_NATIVE_TLS, IPPROTO_TLS_1_2);
+	} else {
+		*sock = socket(bind_addr->sa_family, SOCK_STREAM |
+						     SOCK_NATIVE, IPPROTO_TCP);
+	}
+
+	if (*sock < 0) {
+		LOG_ERR("Failed to create TCP socket: %d", errno);
+		return -errno;
+	}
+
+	if (IS_ENABLED(CONFIG_NET_SOCKETS_SOCKOPT_TLS)) {
+		sec_tag_t sec_tag_list[] = {
+			SERVER_CERTIFICATE_SEC_TAG,
+		};
+
+		ret = setsockopt(*sock, SOL_TLS, TLS_SEC_TAG_LIST,
+				 sec_tag_list, sizeof(sec_tag_list));
+		if (ret < 0) {
+			LOG_ERR("Failed to set TCP secure option %d", errno);
+			ret = -errno;
+		}
+	}
+
+	ret = bind(*sock, bind_addr, bind_addrlen);
+	if (ret < 0) {
+		LOG_ERR("Failed to bind TCP socket %d", errno);
+		return -errno;
+	}
+
+	ret = listen(*sock, MAX_CLIENT_QUEUE);
+	if (ret < 0) {
+		LOG_ERR("Failed to listen on TCP socket %d", errno);
+		ret = -errno;
+	}
+
+	return ret;
+}
+
+static void client_conn_handler(void *ptr1, void *ptr2, void *ptr3)
+{
+	ARG_UNUSED(ptr1);
+	int *sock = ptr2;
+	k_tid_t *in_use = ptr3;
+	int received;
+	int ret;
+	char buf[CONFIG_HTTP_SERVER_SAMPLE_RECEIVE_BUFFER_SIZE];
+	size_t offset = 0;
+	size_t total_received = 0;
+	struct http_req request = {
+		.socket = *sock,
+	};
+
+	http_parser_init(&request.parser, HTTP_REQUEST);
+
+	while (true) {
+		received = recv(request.socket, buf + offset, sizeof(buf) - offset, 0);
+		if (received == 0) {
+			/* Connection closed */
+			LOG_DBG("[%d] Connection closed by peer", request.socket);
+			break;
+		} else if (received < 0) {
+			/* Socket error */
+			ret = -errno;
+			LOG_ERR("[%d] Connection error %d", request.socket, ret);
+			break;
+		}
+
+		/* Parse the received data as HTTP request */
+		(void)http_parser_execute(&request.parser,
+					  &parser_settings, buf + offset, received);
+
+		total_received += received;
+		offset += received;
+
+		if (offset >= sizeof(buf)) {
+			offset = 0;
+		}
+
+		/* If the HTTP request has been completely received, stop receiving data and
+		 * proceed to process the request.
+		 */
+		if (request.received_all) {
+			handle_http_request(&request);
+			break;
+		}
+	};
+
+	(void)close(request.socket);
+
+	*sock = -1;
+	*in_use = NULL;
+}
+
+static int get_free_slot(int *accepted)
+{
+	for (int i = 0; i < MAX_CLIENT_QUEUE; i++) {
+		if (accepted[i] < 0) {
+			return i;
+		}
+	}
+
+	return -1;
+}
+
+static void process_tcp(int *sock, int *accepted)
+{
+	int client;
+	int slot;
+	struct sockaddr_in6 client_addr;
+	socklen_t client_addr_len = sizeof(client_addr);
+	char addr_str[INET6_ADDRSTRLEN];
+
+	client = accept(*sock, (struct sockaddr *)&client_addr,
+			&client_addr_len);
+	if (client < 0) {
+		LOG_ERR("Error in accept %d, try again", -errno);
+		return;
+	}
+
+	slot = get_free_slot(accepted);
+	if (slot < 0 || slot >= MAX_CLIENT_QUEUE) {
+		LOG_ERR("Cannot accept more connections");
+		close(client);
```

Comment:
Ideally, we should return «503 Service Unavailable» to the client.

Reply:
I think ill leave this one as a feature improvement as there is no good clean way of doing it.